File: tests/lockup/utils/assertions.ts
import { PublicKey } from "@solana/web3.js";
import type BN from "bn.js";
import { assert } from "vitest";
import type { ProgramErrorName } from "../../../target/types/sablier_lockup_errors";
import { ProgramErrorCode } from "../../../target/types/sablier_lockup_errors";
import type {
  Amounts,
  LinearTimestamps,
  LinearUnlockAmounts,
  StreamData,
  StreamModel,
} from "../../../target/types/sablier_lockup_structs";
import { getATABalance, getATABalanceMint } from "../../common/anchor-bankrun";
import {
  assertAccountExists,
  assertEqBn,
  assertEqPublicKey,
  assertLteBn,
  expectToThrow as baseExpectToThrow,
} from "../../common/assertions";
import type { LockupTestContext } from "../context";
import type { Stream } from "./types";
import { isLinearModel, isTranchedModel } from "./types";

export function assertEqStreamData(a: StreamData, b: StreamData) {
  assertEqAmounts(a.amounts, b.amounts);

  // Compare model-specific data
  assertEqStreamModel(a.model, b.model);

  assertEqPublicKey(
    a.depositedTokenMint,
    b.depositedTokenMint,
    `Asset mint addresses mismatch: ${a.depositedTokenMint.toBase58()} !== ${b.depositedTokenMint.toBase58()}`,
  );
  assertEqBn(a.salt, b.salt);
  assert.equal(a.isCancelable, b.isCancelable);
  assert.equal(a.isDepleted, b.isDepleted);
  assertEqPublicKey(
    a.sender,
    b.sender,
    `Sender address mismatch: ${a.sender.toBase58()} !== ${b.sender.toBase58()}`,
  );
  assert.equal(a.wasCanceled, b.wasCanceled);
}

export function assertEqLinearTimestamps(a: LinearTimestamps, b: LinearTimestamps) {
  assertEqBn(a.cliff, b.cliff, "cliff timestamps mismatch");
  assertEqBn(a.end, b.end, "end timestamps mismatch");
  assertEqBn(a.start, b.start, "start timestamps mismatch");
}

export function assertEqLinearUnlockAmounts(a: LinearUnlockAmounts, b: LinearUnlockAmounts) {
  assertEqBn(a.cliff, b.cliff, "cliff unlock amounts mismatch");
  assertEqBn(a.start, b.start, "start unlock amounts mismatch");
}

export function expectToThrow(
  promise: Promise<unknown>,
  errorNameOrCode: ProgramErrorName | number,
) {
  return baseExpectToThrow(promise, ProgramErrorCode, errorNameOrCode);
}

export async function assertStreamCreation(
  ctx: LockupTestContext,
  salt: BN,
  beforeCollectionSize: BN,
  beforeSenderTokenBalance: BN,
  expectedStream: Stream,
  recipient: PublicKey = ctx.recipient.keys.publicKey,
) {
  // Assert that core stream accounts exist
  await assertAccountExists(ctx, expectedStream.nftAddress, "Stream NFT doesn't exist");
  await assertAccountExists(ctx, expectedStream.dataAddress, "Stream Data doesn't exist");
  await assertAccountExists(ctx, expectedStream.dataAta, "Stream Data ATA doesn't exist");

  // Assert the contents of the Stream Data account
  const actualStreamData = await ctx.fetchStreamData(salt);
  assertEqStreamData(actualStreamData, expectedStream.data);

  // Fetch the Stream NFT
  const streamNft = await ctx.fetchStreamNft(salt);

  // Assert that the Stream NFT is owned by the recipient
  assertEqPublicKey(
    new PublicKey(streamNft.owner),
    recipient,
    "Stream NFT isn't owned by the recipient",
  );

  // Assert that the Update Authority of the Stream NFT isn't undefined
  if (!streamNft.updateAuthority.address) {
    throw new Error("Stream NFT update authority is undefined");
  }

  // Assert that the Stream NFT has been added to the collection
  assertEqPublicKey(
    new PublicKey(streamNft.updateAuthority.address),
    ctx.nftCollectionAddress,
    "Stream NFT isn't added to the collection",
  );

  // Assert that the collection size has increased by exactly 1
  assertEqBn(
    await ctx.getStreamNftCollectionSize(),
    beforeCollectionSize.addn(1),
    "Collection size should have increased by exactly 1",
  );

  // Assert that the Sender's balance has changed correctly
  const expectedTokenBalance = beforeSenderTokenBalance.sub(expectedStream.data.amounts.deposited);
  const afterSenderTokenBalance = await ctx.getSenderTokenBalance(
    expectedStream.data.depositedTokenMint,
  );
  assertEqBn(expectedTokenBalance, afterSenderTokenBalance, "sender balance not updated correctly");
}

export async function postCancelAssertions(
  ctx: LockupTestContext,
  salt: BN,
  expectedStream: Stream,
  beforeSenderBalance: BN,
) {
  // Assert that the Stream state has been updated correctly
  const actualStreamData = await ctx.fetchStreamData(salt);
  assertEqStreamData(actualStreamData, expectedStream.data);

  // Assert the Sender's ATA balance
  const afterSenderBalance = await getATABalanceMint(
    ctx.banksClient,
    expectedStream.data.sender,
    expectedStream.data.depositedTokenMint,
  );

  const actualBalanceRefunded = afterSenderBalance.sub(beforeSenderBalance);
  assertEqBn(actualBalanceRefunded, expectedStream.data.amounts.refunded);

  // Assert the StreamData ATA balance
  const actualStreamDataBalance = await getATABalanceMint(
    ctx.banksClient,
    expectedStream.dataAddress,
    expectedStream.data.depositedTokenMint,
  );
  const expectedStreamDataBalance = expectedStream.data.amounts.deposited.sub(
    expectedStream.data.amounts.refunded,
  );
  assertEqBn(actualStreamDataBalance, expectedStreamDataBalance);
}

export async function postWithdrawAssertions(
  ctx: LockupTestContext,
  salt: BN,
  txSigner: PublicKey,
  txSignerLamportsBefore: BN,
  treasuryLamportsBefore: BN,
  withdrawalRecipientATA: PublicKey,
  withdrawalRecipientATABalanceBefore: BN,
  expectedStreamData: StreamData,
  streamDataAta: PublicKey,
  streamDataAtaBalanceBefore: BN,
) {
  // Assert that the Stream state has been updated correctly
  const actualStreamData = await ctx.fetchStreamData(salt);
  assertEqStreamData(actualStreamData, expectedStreamData);

  const expectedFee = await ctx.withdrawalFeeInLamports();

  // Get the Lamports balance of the Treasury after the withdrawal
  const treasuryLamportsAfter = await ctx.getTreasuryLamports();

  // Assert that the tx signer lamports balance has decreased by, at least, the withdrawal fee amount.
  const txSignerLamportsAfter = await ctx.getLamportsOf(txSigner);
  assertLteBn(txSignerLamportsAfter, txSignerLamportsBefore.sub(expectedFee));

  // Assert that the Treasury has been credited with the withdrawal fee.
  assertEqBn(treasuryLamportsAfter, treasuryLamportsBefore.add(expectedFee));

  // Get the withdrawal recipient's token balance
  const withdrawalRecipientTokenBalance = await getATABalance(
    ctx.banksClient,
    withdrawalRecipientATA,
  );

  // Assert that the withdrawal recipient's token balance has been changed correctly
  const expectedWithdrawnAmount = expectedStreamData.amounts.withdrawn;
  assertEqBn(
    withdrawalRecipientTokenBalance,
    withdrawalRecipientATABalanceBefore.add(expectedWithdrawnAmount),
  );

  // Assert that the StreamData ATA balance has decreased by the withdrawn amount
  const streamDataAtaBalanceAfter = await getATABalance(ctx.banksClient, streamDataAta);
  assertEqBn(streamDataAtaBalanceAfter, streamDataAtaBalanceBefore.sub(expectedWithdrawnAmount));
}

/* -------------------------------------------------------------------------- */
/*                               INTERNAL LOGIC                               */
/* -------------------------------------------------------------------------- */

function assertEqAmounts(a: Amounts, b: Amounts) {
  assertEqBn(a.deposited, b.deposited, "deposited amounts mismatch");
  assertEqBn(a.refunded, b.refunded, "refunded amounts mismatch");
  assertEqBn(a.withdrawn, b.withdrawn, "withdrawn amounts mismatch");
}

function assertEqStreamModel(a: StreamModel, b: StreamModel) {
  if (isLinearModel(a) && isLinearModel(b)) {
    assertEqLinearTimestamps(a.linear.timestamps, b.linear.timestamps);
    assertEqLinearUnlockAmounts(a.linear.unlockAmounts, b.linear.unlockAmounts);
  } else if (isTranchedModel(a) && isTranchedModel(b)) {
    assertEqBn(
      a.tranched.timestamps.start,
      b.tranched.timestamps.start,
      "tranched start timestamps mismatch",
    );
    assertEqBn(
      a.tranched.timestamps.end,
      b.tranched.timestamps.end,
      "tranched end timestamps mismatch",
    );
    assert.equal(
      a.tranched.tranches.length,
      b.tranched.tranches.length,
      "tranches length mismatch",
    );
    for (let i = 0; i < a.tranched.tranches.length; i++) {
      assertEqBn(
        a.tranched.tranches[i].amount,
        b.tranched.tranches[i].amount,
        `tranche ${i} amount mismatch`,
      );
      assertEqBn(
        a.tranched.tranches[i].timestamp,
        b.tranched.tranches[i].timestamp,
        `tranche ${i} timestamp mismatch`,
      );
    }
  } else {
    assert.fail("Stream models do not match (one is Linear, other is Tranched)");
  }
}

File: trident-tests/fuzz_lockup/helpers/assertions.rs
use crate::helpers::*;
use crate::types::StreamData;
use trident_fuzz::fuzzing::*;

// =============================================================================
// CREATE STREAM ASSERTIONS
// =============================================================================

/// Expected values for stream creation assertions.
pub struct CreateExpected {
    pub salt: u128,
    pub deposit_amount: u64,
    pub start_unlock_amount: u64,
    pub cliff_unlock_amount: u64,
    pub is_cancelable: bool,
}

/// Asserts that all stream accounts were created correctly.
pub fn assert_stream_accounts_created(trident: &mut Trident, accounts: &CreateStreamAccounts) {
    assert!(account_exists(trident, &accounts.stream_nft_mint), "stream_nft_mint account was not created");
    assert!(
        account_exists(trident, &accounts.recipient_stream_nft_ata),
        "recipient_stream_nft_ata account was not created"
    );
    assert!(account_exists(trident, &accounts.stream_data), "stream_data account was not created");
    assert!(account_exists(trident, &accounts.stream_data_ata), "stream_data_ata account was not created");
    assert!(
        account_exists(trident, &accounts.stream_nft_master_edition),
        "stream_nft_master_edition account was not created"
    );
    assert!(account_exists(trident, &accounts.stream_nft_metadata), "stream_nft_metadata account was not created");
}

/// Asserts stream data fields (common to both timestamp and duration variants).
pub fn assert_stream_data_fields(trident: &mut Trident, accounts: &CreateStreamAccounts, expected: &CreateExpected) {
    let stream_data = get_stream_data(trident, &accounts.stream_data);

    // Verify amounts
    assert_eq!(stream_data.amounts.deposited, expected.deposit_amount, "deposit_amount mismatch");
    assert_eq!(stream_data.amounts.start_unlock, expected.start_unlock_amount, "start_unlock_amount mismatch");
    assert_eq!(stream_data.amounts.cliff_unlock, expected.cliff_unlock_amount, "cliff_unlock_amount mismatch");
    assert_eq!(stream_data.amounts.withdrawn, 0, "withdrawn should be 0 initially");
    assert_eq!(stream_data.amounts.refunded, 0, "refunded should be 0 initially");

    // Verify other fields
    assert_eq!(stream_data.salt, expected.salt, "salt mismatch");
    assert_eq!(stream_data.is_cancelable, expected.is_cancelable, "is_cancelable mismatch");
    assert_eq!(stream_data.deposited_token_mint, accounts.deposit_token_mint, "deposit_token_mint mismatch");
    assert_eq!(stream_data.sender, accounts.sender, "sender mismatch");
    assert!(!stream_data.is_depleted, "is_depleted should be false initially");
    assert!(!stream_data.was_canceled, "was_canceled should be false initially");
}

/// Asserts token balances after stream creation.
pub fn assert_stream_token_balances(trident: &mut Trident, accounts: &CreateStreamAccounts, deposit_amount: u64) {
    // Verify stream_data_ata balance
    let stream_data_ata_balance = get_ata_token_balance(trident, &accounts.stream_data_ata);
    assert_eq!(stream_data_ata_balance, deposit_amount, "stream_data_ata balance should equal deposit_amount");

    // Verify stream NFT mint supply
    let stream_nft_mint_supply = get_mint_total_supply(trident, &accounts.stream_nft_mint);
    assert_eq!(stream_nft_mint_supply, 1, "Stream NFT Mint total supply should be 1");

    // Verify recipient received the stream NFT
    let recipient_nft_balance = get_ata_token_balance(trident, &accounts.recipient_stream_nft_ata);
    assert_eq!(recipient_nft_balance, 1, "recipient should have 1 stream NFT");
}

// =============================================================================
// WITHDRAWAL ASSERTIONS
// =============================================================================

/// Asserts withdrawal was executed correctly.
pub fn assert_withdrawal(
    trident: &mut Trident,
    accounts: &WithdrawalAccounts,
    withdraw_amount: u64,
    recipient_ata_balance_before: u64,
    stream_data_ata_balance_before: u64,
) {
    let stream_data = get_stream_data(trident, &accounts.stream_data);

    // Verify withdrawn amount increased
    assert_eq!(stream_data.amounts.withdrawn, withdraw_amount, "withdrawn amount should match withdraw_amount");

    // Verify recipient ATA balance increased
    let recipient_ata_balance_after = get_ata_token_balance(trident, &accounts.withdrawal_recipient_ata);
    assert_eq!(
        recipient_ata_balance_after - recipient_ata_balance_before,
        withdraw_amount,
        "recipient ATA balance should increase by withdraw_amount"
    );

    // Verify stream_data_ata balance decreased
    let stream_data_ata_balance_after = get_ata_token_balance(trident, &accounts.stream_data_ata);
    assert_eq!(
        stream_data_ata_balance_before - stream_data_ata_balance_after,
        withdraw_amount,
        "stream_data_ata balance should decrease by withdraw_amount"
    );

    // Verify is_depleted when fully withdrawn
    assert_depleted_if_complete(&stream_data);
}

/// Asserts is_depleted flag when stream is fully consumed.
fn assert_depleted_if_complete(stream_data: &StreamData) {
    if stream_data.amounts.withdrawn + stream_data.amounts.refunded == stream_data.amounts.deposited {
        assert!(stream_data.is_depleted, "is_depleted should be true when withdrawn + refunded = deposited");
    }
}
